use getDeclaredConstructor().newInstance() in ConfigurationSetup

vorburger opened the pull request:
https://errorprone.info/bugpattern/ClassNewInstance suggests this

and introduce new InvocationTargetError

gsmet commented:
FYI, the CI error was a timeout of our JBoss Nexus. Now the funny thing is that you need to force push somehow to get CI to run again and push the status properly to GitHub.

vorburger commented:
> you need to force push somehow to get CI to run again and push the status properly to GitHub.

done (`git rebase master`)

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java:
```diff
@@ -259,11 +264,7 @@ void finalizeConfigLoader(
-                classConsumer.accept(new GeneratedClassBuildItem(false, name, data));
-            }
-        };
+        final ClassOutput classOutput = (name, data) -> classConsumer.accept(new GeneratedClassBuildItem(false, name, data));
```
So I would have merged this one already... if the "let's add lambdas everywhere" part was out of it.

We are very cautious about where we add lambdas and it requires thinking twice about it.

Could you make your PR only about the proposed change?

vorburger replied:
Sorry, I pre-configure some Auto Save Actions, which caused this. Have just disabled, for future.

dmlloyd commented on core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java:
```diff
@@ -192,11 +192,16 @@ public ConfigurationBuildItem initializeConfiguration(
             throw toError(e);
         } catch (IllegalAccessException e) {
             throw toError(e);
+        } catch (InvocationTargetException e) {
```
This isn't the right handling for this.  The cause of the ITE should be extracted and rethrown, wrapping with `UndeclaredThrowableException` in the event that it's a `Throwable` that isn't a `RuntimeException` or `Error`.

vorburger replied:
OK, see if the new version (force pushed) is better? There is a hiccup though... see the TODO: It's not possible (AFAIK) to re-throw the `InvocationTargetException` as either a `RuntimeException` OR an `Error`, with those `toError`method signatures _returning_ instead of _throwing_. Do you want a PR with that changed? Or just forget about using `ReflectUtil` and inline it for that case,with x2 `throw`?

dmlloyd replied:
Right it should be throwing the *cause*, like this:

```java
// ...
} catch (InvocationTargetException e) {
    try {
        throw e.getCause();
    } catch (RuntimeException | Error e2) {
        throw e2;
    } catch (Throwable t) {
        throw new UndeclaredThrowableException(t);
    }
}
```

It's wordy but it is the "most" correct.

vorburger replied:
done

dmlloyd commented on core/deployment/src/main/java/io/quarkus/deployment/util/InvocationTargetError.java:
```diff
@@ -0,0 +1,9 @@
+package io.quarkus.deployment.util;
+
+public class InvocationTargetError extends InstantiationError {
```
We don't need this; see above comment.

vorburger replied:
sure, it's gone.

gsmet commented:
@vorburger do you plan to update this PR or should we close it?

vorburger commented:
> @vorburger do you plan to update this PR or should we close it?

@gsmet thanks for the ping, I'll try to do this today!

dmlloyd commented:
Could it wait until after #1478 is merged?

vorburger commented:
> Could it wait until after #1478 is merged?

sure thing, I'll hold off... saw that `ConfigurationSetup` significantly changes in #1478, so make sense.

dmlloyd commented:
Thanks!

vorburger committed:
use getDeclaredConstructor().newInstance() in ConfigurationSetup

https://errorprone.info/bugpattern/ClassNewInstance suggests this

using throw e.getCause() instead of instanceof, suggested by @dmlloyd

vorburger commented:
@dmlloyd re-review this?

dmlloyd commented:
Looks OK for me now, @gsmet are you OK with this now?

vorburger commented:
`RestClientITCase` failed in CI, and is reproducible locally with `../../mvnw -Dnative=true clean install` (NB: IT) in `integration-tests/main` (after `../../mvnw clean install` in `core/deployment` on branch with this change)...

```
[ERROR] Failures: 
[ERROR]   RestClientITCase>RestClientTestCase.testMicroprofileCdiClientHeaderPassing:88 1 expectation failed.
JSON path header-name doesn't match.
Expected: some-not-at-all-random-header-value
  Actual: null

[ERROR]   RestClientITCase>RestClientTestCase.testMicroprofileClientHeaderPassing:98 1 expectation failed.
JSON path header-name doesn't match.
Expected: some-not-at-all-random-header-value
  Actual: null
```

vorburger commented:
> `RestClientITCase` failed in CI

This is not related to this change, but a more general problem on current `master` - see #1565.

vorburger commented:
> Looks OK for me now, @gsmet are you OK with this now?

@gsmet merge?

gsmet commented:
Merged, thanks!
